lib/scholar/neighbors/kd_tree.ex

```diff
@@ -0,0 +1,370 @@
+defmodule Scholar.Neighbors.KDTree do
+  @moduledoc """
+  Implements a kd-tree, a space-partitioning data structure for organizing points
+  in a k-dimensional space.
+
+  This is implemented as one-dimensional tensor with indices pointed to highest
+  dimension of the given tensor. Traversal starts by calling `root/0` and then
+  accessing the `left_child/1` and `right_child/1`. The tree is left-balanced.
+
+  Two construction modes are available:
+
+    * `banded/2` - the tensor has min and max values with an amplitude given by `max - min`.
+      It is also guaranteed that the `amplitude * levels(tensor) + 1` does not overflow
+      the tensor. See `amplitude/1` to verify if this holds. This implementation happens
+      fully within `defn`. This version is orders of magnitude faster than the `unbanded/2`
+      one.
+
+    * `unbanded/2` - there are no known bands (min and max values) to the tensor.
+      This implementation is recursive and goes in and out of the `defn`, therefore
+      it cannot be called inside `defn`.
+
+  Each level traverses over the last axis of tensor, the index for a level can be
+  computed as: `rem(level, Nx.axis_size(tensor, -1))`.
+
+  ## References
+
+    * [GPU-friendly, Parallel, and (Almost-)In-Place Construction of Left-Balanced k-d Trees](https://arxiv.org/pdf/2211.00120.pdf).
+  """
+
+  import Nx.Defn
+
+  @derive {Nx.Container, keep: [:levels], containers: [:indexes, :data]}
+  @enforce_keys [:levels, :indexes, :data]
+  defstruct [:levels, :indexes, :data]
+
+  @doc """
+  Builds a KDTree without known min-max bounds.
+
+  If your tensor has a known band (for example, -1 and 1),
```

Comment:
Is this the wording they use in one of the papers? I would expect "bounds" and `bounded/2`.

Reply:
No paper uses it, the idea is that it would reflect "frequency bands" but I think bound and unbound is more common. It may be a bit confusing because they are not really unbound and more of a question if the bounds are known and reasonable.